lib/cocoapods/installer/analyzer.rb

```diff
@@ -249,48 +253,119 @@ def generate_target(target_definition, specs)
           end
         end
 
-        target.pod_targets = generate_pod_targets(target, specs)
-
+        target.pod_targets = pod_targets.select do |pod_target|
+          pod_target.target_definitions.include?(target_definition)
+        end
         target
       end
 
-      # Setup the pod targets for an aggregate target. Group specs and subspecs
-      # by their root to create a {PodTarget} for each spec.
+      # Setup the pod targets for an aggregate target. Deduplicates resulting
+      # targets by grouping by grouping by platform and subspec by their root
+      # to create a {PodTarget} for each spec.
       #
-      # @param  [AggregateTarget] target
-      #         the aggregate target
+      # @param  [Hash{Podfile::TargetDefinition => Array<Specification>}] specs_by_target
+      #         the resolved specifications grouped by target.
       #
-      # @param  [Array<Specification>] specs
-      #         the specifications that need to be installed.
+      # @return [Array<PodTarget>]
+      #
+      def generate_pod_targets(specs_by_target)
+        if config.deduplicate_targets?
+          all_specs = specs_by_target.flat_map do |target_definition, dependent_specs|
+            dependent_specs.group_by(&:root).map do |root_spec, specs|
+              [root_spec, specs, target_definition]
+            end
+          end
+
+          distinct_targets = all_specs.each_with_object({}) do |dependency, hash|
+            root_spec, specs, target_definition = *dependency
+            hash[root_spec] ||= {}
+            (hash[root_spec][[specs, target_definition.platform]] ||= []) << target_definition
+          end
+
+          pod_targets = distinct_targets.flat_map do |_, targets_by_distinctors|
```

Comment:
No skipping the first is something different. This would be perhaps more obvious:
```diff
- distinct_targets.flat_map do |_, targets_by_distinctors|
+ distinct_targets.values.flat_map do |targets_by_distinctors|
```